File: Homework2/IR_HW_2.py
# imports
import numpy as np
import pandas as pd

##################################
# function to create adjacency list
# from graph of the webpages.
##################################
def generate_adjacency_list():
    adjacency_list = dict()

    with open('./graph.txt', 'r') as _file:
        for line in _file:
            connection = line.strip().split()
            if connection[0] not in adjacency_list:
                adjacency_list[connection[0]] = []
                if connection[2] == "1":
                    adjacency_list[connection[0]].append(connection[1])
            else:
                if connection[2] == "1":
                    adjacency_list[connection[0]].append(connection[1])

    return adjacency_list


##################################
# generates M Matrix from the adjacency
# list
##################################
def generate_m_matrix(graph):
    df = pd.DataFrame(columns=graph.keys(), index=graph.keys(), dtype="double")
    df = df.fillna(0.0)

    for key, val in graph.items():
        for i in range(0, len(val)):
            df.loc[val[i], key] = 1 / len(val)

    m_matrix = df.values

    return m_matrix


##################################
# generates original rank vector from
# the adjancecy list
##################################
def generate_original_rank_vector(graph):
    original_rank_vector = []

    for i in range(0, len(graph.keys())):
        original_rank_vector.append(1 / len(graph.keys()))

    original_rank_vector = np.matrix(original_rank_vector).T

    return original_rank_vector


##################################
# generates A matrix using M matrix
# follows the equation given in the
# PageRank presentation.
##################################
def generate_A_matrix(graph, m_matrix, beta):

    teleport_factor = (1 - beta) / len(graph) # this is teleport propbability
    a_matrix = beta * m_matrix + teleport_factor

    return a_matrix


##################################
# power iteration method for M
# matrix using equation r(t+1) = M * r(t)
##################################
def power_converge_m_matrix(rank_vector, m_matrix):
    power = 0
    rank_vector_t = rank_vector
    rank_vector_t_plus_1 = np.empty(shape=(len(rank_vector), 1))
    rank_vector_t_plus_1.fill(0)

    while not np.allclose(rank_vector_t, rank_vector_t_plus_1, rtol=1e-06, atol=1e-06):
        rank_vector_t = rank_vector
        rank_vector_t_plus_1 = m_matrix * rank_vector_t
        rank_vector = rank_vector_t_plus_1
        power += 1

    return power, rank_vector_t_plus_1


##################################
# power iteration method for A
# matrix using equation -
# r(t+1) = Sigma(beta * M Matrix * r(t) + ((1-beta)/(total_nodes)))
##################################
def power_converge_a_matrix(graph, rank_vector, m_matrix, beta):
    power = 0
    rank_vector_t = rank_vector
    rank_vector_t_plus_1 = np.empty(shape=(len(rank_vector), 1))
    rank_vector_t_plus_1.fill(0)
    teleport_factor = (1 - beta) / len(graph) # this is teleport propbability

    while not np.allclose(rank_vector_t, rank_vector_t_plus_1, rtol=1e-06, atol=1e-06):
        rank_vector_t = rank_vector
        rank_vector_t_plus_1 = beta * m_matrix * rank_vector_t + teleport_factor
        rank_vector = rank_vector_t_plus_1
        power += 1

    return power, rank_vector_t_plus_1


if __name__ == '__main__':
    graph = generate_adjacency_list()
    m_matrix = generate_m_matrix(graph)
    rank_vector = generate_original_rank_vector(graph)
    beta = 0.85
    a_matrix = generate_A_matrix(graph, m_matrix, beta)
    m_iterations, final_rank_vector_m = power_converge_m_matrix(rank_vector, m_matrix)
    a_iterations, final_rank_vector_a = power_converge_a_matrix(graph, rank_vector, m_matrix, beta)
    print ('Iterations M : ', m_iterations)
    print ('Iterations A : ', a_iterations)
    print ('Original Rank Vector : ')
    print (rank_vector)
    print ('Matrix M : ')
    print (m_matrix)
    print ('Matrix A : ')
    print (a_matrix)
    print ('Final Rank Vector M : ')
    print (final_rank_vector_m)
    print ('Final Rank Vector A : ')
    print (final_rank_vector_a)

File: Homework2/ir_hw_2.cpp
#include <fstream>
#include <string>
#include <iostream>
#include <map>
#include <vector>
#include <cmath>

using namespace std;

vector<vector<double> > generate_m_matrix(map<char, vector<char> > &adjacencyList){
	vector<vector<double> > m_matrix;
	for(auto i: adjacencyList){
		vector<double> row;
		for(int j = 0; j < adjacencyList.size(); ++j)
			row.push_back(0.0);
		m_matrix.push_back(row);
	}

	for(auto i: adjacencyList){
		for(auto j : i.second){
			m_matrix[j - 'A'][i.first - 'A'] = 1.0 / i.second.size();
		}
	}
	return m_matrix;
}


vector<vector<double> > getOriginalRankVector(map<char, vector<char> > &adjacencyList){
	vector<vector<double> > rank_vector;

	for (int i = 0; i < adjacencyList.size(); ++i)
	{
		vector<double> row;
		row.push_back(1.0/adjacencyList.size());
		rank_vector.push_back(row);
	}

	return rank_vector;
}

vector<vector<double> > generate_a_matrix(vector<vector<double> > &a_matrix, double beta, map<char, vector<char> > &adjacencyList){

	for (int i = 0; i < a_matrix.size(); ++i)
		for (int j = 0; j < a_matrix[i].size(); ++j)
			a_matrix[i][j] = beta * a_matrix[i][j] + ((1.0 - beta) / adjacencyList.size()); 
	return a_matrix;
}


bool checkThreshold(vector<vector<double> > &r_t, vector<vector<double> > &r_t_plus_1){

	for (int i = 0; i < r_t.size(); ++i){
		if(fabs(r_t[i][0] - r_t_plus_1[i][0]) < 1e-06){
			return true;
		}
	}
	return false;
}


void powerConvergeMatrixA(vector<vector<double> > &m_matrix, vector<vector<double> > &original_rank_vector, vector<vector<double> > &converged_vector_a, int &iterations_a){
	int power = 0;

	vector<vector<double> > r_t, r_t_plus_1;
	r_t = original_rank_vector;
	r_t_plus_1 = original_rank_vector;


	for (int i = 0; i < r_t_plus_1.size(); ++i)
		for (int j = 0; j < r_t_plus_1[i].size(); ++j)
			r_t_plus_1[i][j] = 0.0;

	while(!(checkThreshold(r_t, r_t_plus_1))){

		r_t = original_rank_vector;

    	for(int i = 0; i < m_matrix.size(); ++i){
		    for(int k = 0; k < m_matrix.size(); ++k)
		       r_t_plus_1[i][0] += m_matrix[i][k] * r_t[k][0]; 
			r_t_plus_1[i][0] = 0.85 * r_t_plus_1[i][0] + 0.025;
		}
		original_rank_vector = r_t_plus_1;
		power++;
	}
	iterations_a = power;
	converged_vector_a = r_t_plus_1;
	return;
}


void powerConvergeMatrixM(vector<vector<double> > &m_matrix, vector<vector<double> > &original_rank_vector, vector<vector<double> > &converged_vector_m, int &iterations_m){
	int power = 0;

	vector<vector<double> > r_t, r_t_plus_1;
	r_t = original_rank_vector;
	r_t_plus_1 = original_rank_vector;


	for (int i = 0; i < r_t_plus_1.size(); ++i)
			r_t_plus_1[i][0] = 0.0;

	while(!(checkThreshold(r_t, r_t_plus_1))){

		r_t = original_rank_vector;

    	for(int i = 0; i < m_matrix.size(); ++i){
		    for(int k = 0; k < m_matrix.size(); ++k){
		       r_t_plus_1[i][0] += m_matrix[i][k] * r_t[k][0]; 
		   }
		}
		original_rank_vector = r_t_plus_1;

		power++;
	}
	iterations_m = power;
	converged_vector_m = r_t_plus_1;
	return;
}


int main(){
	std::ifstream infile("./graph.txt");
	double beta = 0.85;
	map<char, vector<char> > adjacencyList;
	vector<vector<double> > m_matrix;
	vector<vector<double> > original_rank_vector, converged_vector_m, converged_vector_a;
	int iterations_a, iterations_m;

	if (infile.is_open()) {
	    std::string line;
	    char a, b;
	    int c;
	    while(infile>>a>>b>>c){
	    	if(adjacencyList.find(a) == adjacencyList.end()){
	    		vector<char> links;
	    		adjacencyList[a] = links;
	    		if(c == 1)
	    			adjacencyList[a].push_back(b);
	    	} else{
	    		if(c == 1)
	    			adjacencyList[a].push_back(b);
	    	}
	    }
	    infile.close();
	}

	m_matrix = generate_m_matrix(adjacencyList);

	original_rank_vector = getOriginalRankVector(adjacencyList);

	vector<vector<double> > a_matrix = m_matrix;

	a_matrix = generate_a_matrix(a_matrix, beta, adjacencyList);


	// powerConvergeMatrixA(m_matrix, original_rank_vector, converged_vector_a, iterations_a);
	// powerConvergeMatrixM(m_matrix, original_rank_vector, converged_vector_m, iterations_m);

	cout<<"M Matrix : "<<endl;
	for(auto i: m_matrix){
		for(auto j: i)
			cout<<j<<" ";
		cout<<endl;
	}

	cout<<endl<<"A Matrix : "<<endl;
	for(auto i: a_matrix){
		for(auto j: i)
			cout<<j<<" ";
		cout<<endl;
	}

	cout<<endl<<"Original Rank Vector : "<<endl;
	for(auto i: original_rank_vector){
		for(auto j: i)
			cout<<j<<" ";
		cout<<endl;
	}

	return 0;
}